zeroize/src/lib.rs

```diff
@@ -325,6 +325,32 @@ where
     /// previous reallocations did not leave values on the heap.
     fn zeroize(&mut self) {
         self.iter_mut().zeroize();
+
+        // Zero the capacity of the `Vec` that is not initialized.
+        {
+            // Safety:
+            //
+            // This is safe, because `Vec` never allocates more than `isize::MAX` bytes.
+            // This exact use case is even mentioned in the documentation of `pointer::add`.
+            let extra_capacity_start = unsafe { self.as_mut_ptr().add(self.len()) as *mut u8 };
+            let extra_capacity_len = self.capacity().saturating_sub(self.len());
+
+            for i in 0..(extra_capacity_len * core::mem::size_of::<Z>()) {
```

Comment:
I think there's a simpler option than casting to a `*mut u8`, which is using [`mem::zeroed`](https://doc.rust-lang.org/std/mem/fn.zeroed.html) to produce the all-zero byte pattern, and `ptr::write_volatile` to write it, e.g.

```rust
for i in 0..extra_capacity_len {
    unsafe { ptr::write_volatile(extra_capacity_start.add(i), mem::zeroed(); }
}
```

Reply:
I considered that, but it would result in undefined behavior, when `Z` is not a type where an all-zero pattern is a valid value. However now that I think about it again, it wouldn't make much sense to implement `Zeroize` for such a type. I'll adjust that.

Reply:
Actually on second thought, if it were implemented that way, it would be possible to invoke undefined behavior from safe code. For example this would then result in undefined behavior:

```rust
use core::num::NonZeroU8;
use zeroize::Zeroize;

struct NonZero(NonZeroU8);

impl Zeroize for NonZero {
    fn zeroize(&mut self) {
        self.0 = NonZeroU8::new(1).unwrap();
    }
}

fn main() {
    let mut vec = vec![NonZero(NonZeroU8::new(2).unwrap())];
    vec.clear();
    // undefined behavior: this would create a `NonZeroU8` with a
    // memory-representation of all-zeroes while zeroing the
    // uninitialized memory
    vec.zeroize();
}
```

This may not be a very useful implementation of `Zeroize`, but it would be a way to invoke undefined behavior in safe code. The other version does not have this problem.

Reply:
How would the uninitialized memory be exposed to safe code?

Reply:
> How would the uninitialized memory be exposed to safe code?

It doesn't need to. According to the [doc of `mem::zeroed`](https://doc.rust-lang.org/std/mem/fn.zeroed.html):

> There is no guarantee that an all-zero byte-pattern represents a valid value of some type `T`. For example, the all-zero byte-pattern is not a valid value for reference types (`&T` and `&mut T`). Using `zeroed` on such types causes immediate [undefined behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) because [the Rust compiler assumes](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initialization-invariant) that there always is a valid value in a variable it considers initialized.

Which links to a [section in the docs about undefined behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html):

> It is the programmer's responsibility when writing `unsafe` code to ensure that any safe code interacting with the `unsafe` code cannot trigger these behaviors. `unsafe` code that satisfies this property for any safe client is called sound; if `unsafe` code can be misused by safe code to exhibit undefined behavior, it is unsound.

The same page also states that

> Rust code is incorrect if it exhibits any of the behaviors in the following list.
> * [...]
> * Invalid values for a type with a custom definition of invalid values. In the standard library, this affects [`NonNull<T>`](https://doc.rust-lang.org/core/ptr/struct.NonNull.html) and [`NonZero*`](https://doc.rust-lang.org/core/num/index.html).

The example above would therefore be both unsound and incorrect, which is possible from safe code. The values do not need to be accessible by safe code to create the problem here.

With the method which zeroes the memory using `u8` this would not be the case, as `u8` is guaranteed to be able to hold a value of `0`.

Reply:
> With the method which zeroes the memory using u8 this would not be the case, as u8 is guaranteed to be able to hold a value of 0.

After a pointer cast from the original type. It's six of one, half dozen of another... in the end you're writing the zero byte pattern either way.

Reply:
> How would the uninitialized memory be exposed to safe code?

The uninitialized guarantee is not one that must only for safe code, but for all code. The purpose of `unsafe` code is only to allow operations where the compiler can't prove them sound but you can. It does not allow violating the validity guarantees that the languages requires for its value semantics.

The impl for `Vec<_>` does not require any value of `Z` to actually exist and it is unsound to generally create one from thin air. (If `Zeroize` were an `unsafe trait` then *it* could require that all implementors are valid for the all-zero pattern and then and only then `mem::zeroed` would be sound).

That shows an interesting alternative though: if there are any values in the initialized part of the vector then one could copy them instead of zeroing the bytes.

Reply:
In what way is the excess capacity of a `Vec` not uninitialized memory by design?

What ever reads from it?

What does a normal `Vec::with_capacity` call initialize it to?

Reply:
It's not about the excess capacity of `Vec`, that's fine and you can zero it nothing creates a reference to it nor do copy its bytes by value. It is only observed through raw pointers. But any values on your own function stack need to be initialized.

> How is casting *mut T to *mut u8 and writing core::mem::size_of::<T> zero-bytes to it any different from writing one mem::zeroed<T>() to it?

The latter creates a (temporary but still) value of type `T`. And this *value* needs to be valid according to Rust's type system.

Reply:
The subexpression `core::mem::zeroed<NonNull<T>>()` is UB. Period. If your program executes it, it is game over.

Reply:
> The latter creates a (temporary but still) value of type T. 

What concretely could go wrong with this line:

```rust
unsafe { ptr::write_volatile(extra_capacity_start.add(i), mem::zeroed()); }
```

What is a better solution?

Reply:
It contains the subexpression `mem::zeroed()`. And as said before, it is executed and that is UB for certain types.

Reply:
Can you answer this question?

> What is a better solution?

But also, it's UB because it's writing the zero byte pattern to memory, is it not?

Is that not *exactly* what casting to `*mut u8` and looping over `core::mem::size_of::<T>` would do?

Reply:
The problem in this case is that the type `T` for `mem::zeroed()` can be arbitrarily changed by safe code and the only invariant that needs to be upheld for `mem::zeroed()` is that `T` is a type, for which an all-zeroes byte pattern is a valid value. Safe code can break that invariant, which makes the `unsafe` code unsound.

Memory behind a raw pointer can hold any value without any problem, because Rust does not consider it to be initialized. The problem only arises when an invalid value of type `T` exists, which Rust considers initialized, which `mem::zeroed<T>()` does.

Looping over all the bytes in the capacity and writing is not UB, because they all exist behind a raw pointer. The value which is actually written to the pointed address, is a valid `u8`, so no invalid value of any type ever exists. Casting a pointer to a pointer of a different type is not UB, as long as the pointer is used correctly (e.g. not using it while it aliases).

The current implementation should be free of undefined behavior.

> I feel there's an excess of waxing philosophical going on here...

The problem with undefined behavior is that the code may work perfectly today, but tomorrow another optimization is implemented in the Rust compiler and the code may break, because Rust does not guarantee that it works.

Reply:
> What is a better solution?

The implementation, as currently written.

Reply:
It still seems like 6 of one, half dozen of another to me, except looping a byte-at-a-time and doing pointer arithmetic around `core::mem::zeroed::<NonNull<T>>()` is more complex and potentially slower.

It seems the crux of this is...

> Memory behind a raw pointer can hold any value without any problem, because Rust does not consider it to be initialized. The problem only arises when an invalid value of type T exists, which Rust considers initialized, which mem::zeroed<T>() does.

...but we're talking about a buffer which is defined as being:

> What does a normal Vec::with_capacity call initialize it to?
>
> Nothing, it does not create a value of type T though and just leaves the memory completely uninitialized.

The contract of `Vec` is always to initialize this capacity in some way before reading from it.

It seems the claim is it might be considered initialized by the Rust compiler when it is uninitialized, but before its API permits any reads, it will be initialized again.

Can either of you give a concrete example of a tractable problem which is possible given the contract of `Vec` to always (re)initialize this memory which would not occur with pointer-based zeroing?

Reply:
> The implementation, as currently written in this PR.

I can merge this as is, but it's really making me want to have a stable `memset` intrinsic (e.g. `volatile_set_memory`)

Reply:
What you are doing to the excess buffer of the vector is and never was the problem. You are indeed free to do *with the extra capacity* of bytes whatever you want. But not *however* you want. The code that writes invalid or uninitialized values into it must still be UB-free and executing `mem::zeroed::<T>` is **not**.

Reply:
> Can either of you give a concrete example of a tractable problem which is possible given the contract of Vec to always (re)initialize this memory which would not occur with pointer-based zeroing?

The problem is not the `Vec` or its memory, the problem is simply that the code could call `mem::zeroed<NonNull<()>>`, which is UB. The compiler might make optimizations that would crash the program if such a value ever exists. This would be the same problem in every program, even without a `Vec` being involved.

Reply:
Okay, I think I get it now. Thank you for the explanation.